src/requests/sign-request.ts

```diff
@@ -59,51 +60,58 @@ const getSortedAndSignedHeaders = (headers: Record<string, string>, timestamp: n
   return { sortedHeaders, signedHeaders }
 }
 
-const getSubjectHeaders = (contextHeaders: any): SubjectHeader => {
-  let headers: SubjectHeader = {}
-
-  if (contextHeaders[ContentfulUserIdHeader]) {
-    headers[ContentfulUserIdHeader] = contextHeaders[ContentfulUserIdHeader]
-  } else if (contextHeaders[ContentfulAppIdHeader]) {
-    headers[ContentfulAppIdHeader] = contextHeaders[ContentfulAppIdHeader]
-  }
-
-  return headers
-}
-
 /**
  * Given a secret, a canonical request, a timestamp and context headers, generates a signature.
  * ~~~
  * @category Requests
  */
-export const signRequest = (
+// Remove when this eslint rule covers all the cases
+// https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/ROADMAP.md
+/*eslint-disable no-unused-vars, no-redeclare*/
+export function signRequest(
+  rawSecret: Secret,
+  rawCanonicalRequest: CanonicalRequest,
+  rawTimestamp?: Timestamp
+): SignedRequestWithoutContextHeaders
+export function signRequest(
+  rawSecret: Secret,
+  rawCanonicalRequest: CanonicalRequest,
+  rawTimestamp?: Timestamp,
+  rawContext?: Context<SubjectHeadersApp>
+): SignedRequestWithContextHeadersWithApp
+export function signRequest(
   rawSecret: Secret,
   rawCanonicalRequest: CanonicalRequest,
-  rawTimestamp: Timestamp = Date.now(),
-  rawContextHeaders: ContextHeaders = {} as ContextHeaders
-): SignedRequestHeaders => {
+  rawTimestamp?: Timestamp,
+  rawContext?: Context<SubjectHeadersUser>
+): SignedRequestWithContextHeadersWithUser
+export function signRequest(
+  rawSecret: Secret,
+  rawCanonicalRequest: CanonicalRequest,
+  rawTimestamp?: Timestamp,
+  rawContext?: any
+) {
+  const maybeDefaultTimestamp = rawTimestamp ?? Date.now()
```

Comment:
nit: I don't really like that variable name. It doesn't really say what it is actually storing. Also: Why not using the fallback date in the function's signature as we did in the old impementation?